mypy/typeanal.py

```diff
@@ -374,6 +374,12 @@ def try_analyze_special_unbound_type(self, t: UnboundType, fullname: str) -> Opt
                 self.fail("NotRequired[] must have exactly one type argument", t)
                 return AnyType(TypeOfAny.from_error)
             return RequiredType(self.anal_type(t.args[0]), required=False)
+        elif fullname in ('typing_extensions.Self', 'typing.Self'):
+            try:
+                bound = self.named_type(self.api.type.fullname)  # type: ignore
```

Comment:
Should do the appropriate `isinstance` checks instead, that's cleaner and probably makes mypyc happier